crit/mempages.go

```diff
@@ -0,0 +1,134 @@
+package crit
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/checkpoint-restore/go-criu/v6/crit/images/pagemap"
+)
+
+const pageSize = 4096
+
+// MemoryReader is a struct used to retrieve
+// the content of memory associated with a specific process ID (pid).
+// New instances should be created with NewMemoryReader()
+type MemoryReader struct {
+	checkpointDir  string
+	pid            int
+	pagesID        uint32
+	pagemapEntries []*pagemap.PagemapEntry
+}
+
+// NewMemoryReader creates a new MemoryReader instance with all fields populated.
+func NewMemoryReader(checkpointDir string, pid int) (*MemoryReader, error) {
+	pagemapImg, err := getImg(filepath.Join(checkpointDir, fmt.Sprintf("pagemap-%d.img", pid)), &pagemap.PagemapHead{})
+	if err != nil {
+		return nil, err
+	}
+
+	pagesID := pagemapImg.Entries[0].Message.(*pagemap.PagemapHead).GetPagesId()
+
+	pagemapEntries := make([]*pagemap.PagemapEntry, 0)
+
+	for _, entry := range pagemapImg.Entries[1:] {
+		pagemapEntries = append(pagemapEntries, entry.Message.(*pagemap.PagemapEntry))
+	}
+
+	return &MemoryReader{
+		checkpointDir:  checkpointDir,
+		pid:            pid,
+		pagesID:        pagesID,
+		pagemapEntries: pagemapEntries,
+	}, nil
+}
+
+// GetMemPages retrieves content of memory associated with a pid.
+func (mr *MemoryReader) GetMemPages(start, end uint64) (*bytes.Buffer, error) {
+	// Calculate and check the size of the memory area
+	size := end - start
+	if size == 0 {
+		return nil, fmt.Errorf("memory size must be greater than 0")
+	}
+
+	startPage := start / pageSize
+	endPage := end / pageSize
+
+	var buffer bytes.Buffer
+
+	for pageNumber := startPage; pageNumber <= endPage; pageNumber++ {
+		var page []byte = nil
+
+		pageMem, err := mr.getPage(pageNumber)
+		if err != nil {
+			return nil, err
+		}
+
+		if pageMem != nil {
+			page = pageMem
+		} else {
+			page = bytes.Repeat([]byte("\x00"), int(pageSize))
+		}
+
+		var nSkip, nRead uint64
+
+		if pageNumber == startPage {
+			nSkip = start - pageNumber*pageSize
+			if startPage == endPage {
+				nRead = size
+			} else {
+				nRead = pageSize - nSkip
+			}
+		} else if pageNumber == endPage {
+			nSkip = 0
+			nRead = end - pageNumber*pageSize
+		} else {
+			nSkip = 0
+			nRead = pageSize
+		}
+
+		if _, err := buffer.Write(page[nSkip : nSkip+nRead]); err != nil {
+			return nil, err
+		}
+	}
+
+	return &buffer, nil
+}
+
+// getPage retrieves a memory page from the pages.img file.
+func (mr *MemoryReader) getPage(pageNo uint64) ([]byte, error) {
+	var offset uint64 = 0
+
+	// Iterate over pagemap entries to find the corresponding page
+	for _, m := range mr.pagemapEntries {
+		found := false
+		for i := 0; i < int(*m.NrPages); i++ {
+			if m.GetVaddr()+uint64(i)*pageSize == pageNo*pageSize {
+				found = true
+				break
+			}
+			offset += pageSize
+		}
+
+		if !found {
+			continue
+		}
+
+		f, err := os.Open(filepath.Join(mr.checkpointDir, fmt.Sprintf("pages-%d.img", mr.pagesID)))
+		if err != nil {
```

Comment:
Can you trigger this error cases by temporarily renaming the expected `pages-%d.img` to not exist.